### PR TITLE
Improving train time perf

### PR DESCRIPTION
**Pull request by gdadlaney**

### CODE REVIEW

File: lib/python/examples/fwdllm/trainer/forward_training/tc_transformer_trainer_distribute.py [hunk 1]
```diff
@@ -348,132 +407,138 @@ def _prepare_perturbation_tensors(self, device, v_buffer):
...
-
-                    loss, jvp = self._compute_forward_jvp(device, x, labels, v_params)
-
-                    self._accumulate_and_extract_grads(device, jvp, v_params)
+                    loss = self._train_one_batch(device, batch, epoch, batch_idx, v_buffer)
 
-                    current_loss = loss.item()
-                    tr_loss += current_loss
+                    # Optimization: Keep tr_loss on device.
+                    tr_loss += loss
```
Comment (AishwwaryaM) [inline lib/python/examples/fwdllm/trainer/forward_training/tc_transformer_trainer_distribute.py]: If this is not used as we saw before, we can just remove this altogether?

**Review by AishwwaryaM — DISMISSED** [review]

File: lib/python/examples/fwdllm/expts/run_tc_expts/json_scripts/trainer_0.json
Comment (AishwwaryaM) [inline lib/python/examples/fwdllm/expts/run_tc_expts/json_scripts/trainer_0.json]: can you update this as well to the final value that you decided was best?

File: lib/python/flame/mode/horizontal/syncfl/fwdllm_trainer.py [hunk 2]
```diff
@@ -547,6 +547,12 @@ def _send_grads(self, tag: str) -> None:
 
         channel._selector._cleanup_send_ends()
 
+        # Optimization: Perform GC and CUDA memory cleanup after sending gradients.
+        # This moves the "stop the world" synchronous flushes out of the measured 
+        # training/evaluation phases and into the idle time between rounds.
+        gc.collect()
```
Comment (dhruvsgarg) [inline lib/python/flame/mode/horizontal/syncfl/fwdllm_trainer.py:553]: So besides the training_delay_s and a few seconds in cuda-cleanup there were no other overheads in the trainer code? Just confirming.

File: lib/python/flame/mode/horizontal/syncfl/fwdllm_aggregator.py [hunk 3]
```diff
@@ -1107,11 +1102,6 @@ def _prepare_distribution_payload(self, task_to_perform: str):
             MessageType.ITERATION_PER_DATA_ID: self.iteration_per_data_id,
         }
 
-        if self._is_model_updated:
```
Comment (dhruvsgarg) [inline lib/python/flame/mode/horizontal/syncfl/fwdllm_aggregator.py:1110]: This state cleanup seems important. Why is it being removed?

Comment (dhruvsgarg) [reply lib/python/flame/mode/horizontal/syncfl/fwdllm_aggregator.py:1110]: Okay its happening elsewhere, and git diff on web ui is messing it up

File: lib/python/flame/mode/horizontal/syncfl/fwdllm_aggregator.py [hunk 4]
```diff
@@ -854,7 +850,6 @@ def _aggregate_grads_sync(self, tag: str) -> None:
         self._process_aggregation_goal_met(tag, channel, is_async=False)
         self.log_memory("end _aggregate_grads_sync", self.device)
 
-
     @timer_decorator
     def _force_cuda_memory_cleanup(self):
```
Comment (dhruvsgarg) [inline lib/python/flame/mode/horizontal/syncfl/fwdllm_aggregator.py:858]: Seems that cuda memory cleanup and GC only happens when eval is done, which means once per round. This is definitely okay but im just thinking about the case where the num-iterations go up for further data bins. So three questions: (i) is the current frequency of gc enough, (ii) how much memory gets freed per gc approx, is it any significant? and (iii) if it is significant, are there any other places in the aggregator where we can do this? Cause var check is happening often, right?

File: lib/python/examples/fwdllm/trainer/forward_training/tc_transformer_trainer_distribute.py (hunk 1 above)
Comment (dhruvsgarg) [reply lib/python/examples/fwdllm/trainer/forward_training/tc_transformer_trainer_distribute.py]: the `tr_loss` is being returned at the bottom, but I guess you're saying that we dont use it at the trainer nor do we use it for stat utility?

Comment (dhruvsgarg) [reply lib/python/examples/fwdllm/trainer/forward_training/tc_transformer_trainer_distribute.py]: Btw gaurav, did you incorporate the change for the corrected stat utility return value from the branch i had pointed you to? I dont see that on the trainer side changes

File: lib/python/examples/fwdllm/trainer/forward_training/tc_transformer_trainer_distribute.py [hunk 5]
```diff
@@ -348,132 +407,138 @@ def _prepare_perturbation_tensors(self, device, v_buffer):
...
+
+            with torch.no_grad():
+                pred = self.model(x)
+                if hasattr(pred, "logits"):
+                    logits = pred.logits
+                elif isinstance(pred, (tuple, list)):
+                    logits = pred[0]
+                else:
+                    logits = pred
+                loss = self.base_trainer.oort_loss(logits, labels.view(-1), epoch=0, batch_idx=0, reduction="mean")
```
Comment (dhruvsgarg) [inline lib/python/examples/fwdllm/trainer/forward_training/tc_transformer_trainer_distribute.py:392]: It seems to continue to use stat_utility just from a single data-bin given x and labels would be that? You can put a print for even one iteration where the trainer prints out loss values for different data bins for calculating stat util.

File: lib/python/examples/fwdllm/expts/run_tc_expts/json_scripts/trainer_0.json
Comment (dhruvsgarg) [reply lib/python/examples/fwdllm/expts/run_tc_expts/json_scripts/trainer_0.json]: I agree, i think we are anyway not doing dynamic sequence length right now. So we can use the P99 values for the trainer and aggregator based on Gaurav's analysis from previous time?

File: lib/python/examples/fwdllm/expts/run_tc_expts/json_scripts/aggregator.json [hunk 6]
```diff
@@ -50,9 +50,9 @@
         "comm_round": 3000,
         "ci": 0,
         "dataset": "agnews",
-        "data_file_path": "/home/dgarg39/$FWDLLM_USER/fednlp_data/data_files/agnews_data.h5",
-        "partition_file_path": "/home/dgarg39/$FWDLLM_USER/fednlp_data/partition_files/agnews_partition.h5",
-        "partition_method": "uniform",
+        "data_file_path": "/coc/scratch/dgarg/fl_datasets/fwdllm/fednlp_data/data_files/agnews_data.h5",
+        "partition_file_path": "/coc/scratch/dgarg/fl_datasets/fwdllm/fednlp_data/partition_files/agnews_partition.h5",
+        "partition_method": "niid_label_clients=100_alpha=100",
```
Comment (dhruvsgarg) [inline lib/python/examples/fwdllm/expts/run_tc_expts/json_scripts/aggregator.json:55]: This file should also have some sequence length config values right?

**Review by dhruvsgarg — COMMENTED** [review]
Please take a look at the few comments

File: lib/python/examples/fwdllm/expts/run_tc_expts/json_scripts/trainer_0.json
Comment (gdadlaney) [reply lib/python/examples/fwdllm/expts/run_tc_expts/json_scripts/trainer_0.json]: Done, updated to 192. Also, verified that the sequence lengths don't change with the distribution of the data.

File: lib/python/flame/mode/horizontal/syncfl/fwdllm_aggregator.py (hunk 3 above)
Comment (gdadlaney) [reply lib/python/flame/mode/horizontal/syncfl/fwdllm_aggregator.py:1110]: Yeah, this is moved to a common method now: _update_state_after_payload_prepared()

File: lib/python/examples/fwdllm/expts/run_tc_expts/json_scripts/aggregator.json (hunk 6 above)
Comment (gdadlaney) [reply lib/python/examples/fwdllm/expts/run_tc_expts/json_scripts/aggregator.json:55]: Verified that the p99 of sequence lengths don't change with the distribution of the data.

File: lib/python/examples/fwdllm/trainer/forward_training/tc_transformer_trainer_distribute.py (hunk 1 above)
Comment (gdadlaney) [reply lib/python/examples/fwdllm/trainer/forward_training/tc_transformer_trainer_distribute.py]: Cleaned up `tr_loss` and `avg_loss` (derived) that were not being used anywhere to avoid confusion.

@dhruvsgarg, will incorporate stat utility changes in my next PR

File: lib/python/examples/fwdllm/trainer/forward_training/tc_transformer_trainer_distribute.py (hunk 5 above)
Comment (gdadlaney) [reply lib/python/examples/fwdllm/trainer/forward_training/tc_transformer_trainer_distribute.py:392]: Current code for stat_utility calculation is another example of spaghetti code. I'll refactor this in the stat_utility calculation PR

File: lib/python/flame/mode/horizontal/syncfl/fwdllm_aggregator.py (hunk 4 above)
Comment (gdadlaney) [reply lib/python/flame/mode/horizontal/syncfl/fwdllm_aggregator.py:858]: i) Yeah, clean-up only after eval makes sense. GPU mem utilization only spikes during forward passes
ii) Yes, 1160 MB after every eval
iii) I doubt var check would spike GPU mem utilization

Eval mem util logs:
1. [MEM:start eval_model] GPU Allocated: 31.99 MB | GPU Reserved: 614.47 MB
2. [MEM:start compute_metrics] GPU Allocated: 569.86 MB | GPU Reserved: 1774.19 MB  **(After Eval)**
3. Runtime of _force_cuda_memory_cleanup: 0.346717s (Round=1, DataId=147, Iter=52, TrainerId=None)
4. [MEM:end eval_model] GPU Allocated: 569.86 MB | GPU Reserved: 614.47 MB

File: lib/python/flame/mode/horizontal/syncfl/fwdllm_trainer.py (hunk 2 above)
Comment (gdadlaney) [reply lib/python/flame/mode/horizontal/syncfl/fwdllm_trainer.py:553]: Not really, These are the list of optimizations:
1. Buffer creation optimizations
2. Autocasting from FP32 to F16 to use the hardware acceleration on TensorCores in A40s
3. Decreasing max_sequence_length from 256 to 192
4. Avoided GPU to CPU data movement in one instance
5. Cuda_cleanup & GC after sending gradients

File: lib/python/flame/mode/horizontal/syncfl/fwdllm_aggregator.py (hunk 4 above)
Comment (dhruvsgarg) [reply lib/python/flame/mode/horizontal/syncfl/fwdllm_aggregator.py:858]: Okay nice. Quick clarification again since I think I mis-wrote earlier. An eval must be happening at each iteration right (since the trainer needs to send the stat utility)? Ive written "once per round" above which is incorrect, so just wanted to clarify since you didnt raise a correction for it either

Comment (gdadlaney) [reply lib/python/flame/mode/horizontal/syncfl/fwdllm_aggregator.py:858]: No, actually here we're only looking at gc clean up on the Aggregator, where eval only happens only when the var threshold is reached (Once per data bin) <br><br>

Yes, on the trainer, there are 3 forward passes happening per iteration. 2 for the JVP calculation & one for stat utility calculation. And yes, we're doing one GC clean-up after sending the gradients over the channel after every iteration

**Review by dhruvsgarg — APPROVED** [review]
LGTM
